[red-knot] Add notebook support

MichaReiser opened the pull request:
## Summary

This PR adds notebook support to red knot. 

The way notebooks are handled in Ruff is that it parses the notebook and concatenates the content of all cells to a single string. The concatenated string is then passed to the linter or formatter. Reported diagnostics are mapped back to their origin cells by using a source map. 

This PR doesn't implement any of the fancy re-mapping logic but adds support for reading notebooks into a structured representation. This requires two additions:

* A new `System::read_to_notebook` method that reads a system path to a `Notebook`. The motivation for this method is that the LSP uses a structured notebook representation (slightly different from the representation used by `ruff_notebook`) that's cheaper to convert to a `ruff_notebook::Notebook` then reparsing the notebook from source.
* I renamed `SourceCode` to `SourceFile` and changed it to use an internal enum that is either `Text` or `Notebook` to account for the fact that a file can either be a text or notebook.


## Open questions

I'm not too happy with the name `SourceFile`. I'm open for better suggestions but I also don't want to block the PR on finding the perfect name (the query is used very sparsely and a rename is quickly done)


## Cell and embedded file support

I intentionally excluded support for cell-based queries or a design for supporting embedded files (e.g. a doctest block in a python file). I think we could support both and a way of doing it could be to:

* Make `SourceFile` a salsa tracked struct
* Add additional `Cell` and `Embedded` variants that either store another salsa tracked struct, or at least store the origin file id and an index to identify the cell or embed in the file. 

Methods that can operate on any source-file would be changed to accept a `SourceFile` instead of a `File` argument. However, it's unclear if we want this:

* The only use case for a cell-level granular query today is to format a single cell. Formatting a single cell is so fast that it doesn't justifies creating a query for it. 
* It's unclear if we have many queries that can operate on any file content or if this is more a one-off, in which case `Cell` and `Embed` could be salsa tracked structs and we define `format_cell` and `parse_cell` etc. queries rather than having one generic `parse` query.
* An alternative is to introduce a new salsa tracked on top of `SourceFile` with the variants `File`, `Embed`, and `Cell`. 

## Test Plan

I added a notebook file to my test folder and verified that Red Knot runs the lint rules on its content.

MichaReiser commented:
> This is definitely the biggest issue I have with the PR; I also don't much like SourceFile. Perhaps FileContents?

I strongly prefer a less generic name because I think this API shouldn't be used to read any non-Python files (or at least any non-source files). For example, if we consider using the' File' API for this in the future, we should use `File.read_to_string` to read `pth` files.

AlexWaygood commented:
> I strongly prefer a less generic name because I think this API shouldn't be used to read any non-Python files (or at least any non-source files). For example, if we consider using the' File' API for this in the future, we should use `File.read_to_string` to read `pth` files.

I see. This changes my understanding of that function, since:
- My understanding of the `File` struct from the name of the structs and the comments in the source code is that it is an abstract representation of an arbitrary file: not necessarily a Python file.
- The `source_text()` function accepts an arbitrary `File` object
- The `source_text()` function was previously documented as simply being a function that simply "reads the contents of a file".

If `source_text()`/`source_file()` is only meant to be used to read files that we expect to contain Python source code, I think that should be documented better in the docstring for that function.

If your concern is that `FileContents` would be too generic as a name, maybe `SourceFileContents`? I also don't mind `SourceCode` too much, even with this change! Agree it's slightly awkard, but it feels more expressive to me than `SourceFile`. Fundamentally it feels to me like this struct represents what the file _contains_ rather than the file itself. You cannot retrieve the file's metadata, its path, etc., from this struct: only the contents of the file.

AlexWaygood commented:
Note that we've already been using (for a few weeks) `source_text()` to read the source of a file that doesn't contain Python source code: typeshed's `VERSIONS` file: https://github.com/astral-sh/ruff/blob/9a2dafb43d36b62d45a604dea58224a0884cd6e5/crates/red_knot_module_resolver/src/typeshed/versions.rs#L72-L79

carljm reviewed (approved):
This looks reasonable to me!

I don't have strong feelings about the naming.

dhruvmanila commented on crates/ruff_db/src/source.rs:
```diff
@@ -1,74 +1,139 @@
+    kind: SourceFileKind,
+}
+
+#[derive(Eq, PartialEq)]
+enum SourceFileKind {
+    Text(String),
+    Notebook(Notebook),
+}
```
Ignore if this isn't relevant but I was wondering if there's any reason for not using the `SourceKind` or borrowing the name:

https://github.com/astral-sh/ruff/blob/9a2dafb43d36b62d45a604dea58224a0884cd6e5/crates/ruff_linter/src/source_kind.rs#L19-L25

MichaReiser replied:
I wasn't aware of it. There are too many `PySourceType` and kind types haha. 

The only reason right now is that it's in `ruff_linter` and we can't provide all it's API (e.g. we don't want `from_path`)

dhruvmanila commented on crates/ruff_notebook/src/notebook.rs:
```diff
@@ -205,6 +205,22 @@ impl Notebook {
+    /// Creates an empty notebook.
+    ///
+    ///
+    pub fn empty() -> Self {
+        Self::from_raw_notebook(
+            RawNotebook {
+                cells: vec![],
+                metadata: RawNotebookMetadata::default(),
+                nbformat: 4,
+                nbformat_minor: 5,
+            },
+            false,
+        )
+        .unwrap()
+    }
```
Technically, an empty notebook would contain at least a single empty cell. I verified that by creating an empty notebook in VS Code and Jupyter Lab. The use-case for this seems to be only when there's some error when creating a Ruff notebook in Red Knot. I think it make sense to use this representation (one empty code cell) for an empty notebook.

MichaReiser commented on crates/ruff_db/src/source.rs:
```diff
@@ -1,74 +1,139 @@
+    let _span = tracing::trace_span!("source_file", ?file).entered();
+
+    if let Some(path) = file.path(db).as_system_path() {
+        if path.extension().is_some_and(|extension| {
+            PySourceType::try_from_extension(extension) == Some(PySourceType::Ipynb)
+        }) {
```
Yes, that would be possible. But I'm going to introduce the API anyway and I don't like the `PySourceType` default fallback.

MichaReiser commented:
I reverted the name back to `SourceCode`, changed `versions` to use `file.read_to_string`, and ensured that `Notebook::empty` contains an empty code cell.
